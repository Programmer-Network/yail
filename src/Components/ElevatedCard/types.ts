import { ReactNode } from "react";

import { IIconProps } from "../Icon/types";

export interface IElevatedCardProps {
  title: string;
  description?: string;
  icon?: IIconProps;
  children?: ReactNode;
  className?: string;
  onClick?: () => void;
  dataTestId?: string;
}
