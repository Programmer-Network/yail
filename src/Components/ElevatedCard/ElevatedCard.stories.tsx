import { action } from "storybook/actions";

import ElevatedCard from ".";
import Button from "../Button";
import { ButtonVariantEnum } from "../Button/types";

export default {
  title: "Components/ElevatedCard",
  component: ElevatedCard,
  tags: ["autodocs"]
};

export const Basic = () => (
  <ElevatedCard
    title='Curated Resources'
    description='Explore our collection of tools, libraries, and tutorials to enhance your development workflow.'
  />
);

export const WithIcon = () => (
  <ElevatedCard
    title='Curated Resources'
    description='Explore our collection of tools, libraries, and tutorials to enhance your development workflow.'
    icon={{
      iconName: "IconBook",
      className: "w-6 h-6 text-primary"
    }}
  />
);

export const WithChildren = () => (
  <ElevatedCard
    title='Curated Resources'
    description='Explore our collection of tools, libraries, and tutorials to enhance your development workflow.'
    icon={{
      iconName: "IconAddYoutube",
      className: "w-6 h-6 text-primary"
    }}
  >
    <div className='mt-4 flex gap-2'>
      <Button>Explore</Button>
      <Button variant={ButtonVariantEnum.SECONDARY}>Learn More</Button>
    </div>
  </ElevatedCard>
);

export const Clickable = () => (
  <ElevatedCard
    title='Curated Resources'
    description='Explore our collection of tools, libraries, and tutorials to enhance your development workflow.'
    icon={{
      iconName: "IconGroupWork",
      className: "w-6 h-6 text-primary"
    }}
    onClick={() => action("Card clicked")()}
  />
);

export const CustomStyling = () => (
  <ElevatedCard
    title='Curated Resources'
    description='Explore our collection of tools, libraries, and tutorials to enhance your development workflow.'
    icon={{
      iconName: "IconSkull",
      className: "w-6 h-6 text-primary"
    }}
    className='bg-primary-50 border-primary-100 max-w-sm border'
  />
);
