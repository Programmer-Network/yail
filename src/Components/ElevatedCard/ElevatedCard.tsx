import classNames from "classnames";
import React from "react";

import { Icon } from "../Icon";
import { IIconProps } from "../Icon/types";
import { H5, Paragraph } from "../Typography";
import { IElevatedCardProps } from "./types";

const ElevatedCard: React.FC<IElevatedCardProps> = ({
  title,
  description,
  icon,
  children,
  className,
  onClick,
  dataTestId = "elevated-card"
}) => {
  return (
    <div
      data-testid={dataTestId}
      className={classNames(
        "border-border bg-text/1 flex flex-col rounded-lg border p-6",
        {
          "hover:bg-text/3 cursor-pointer": onClick
        },
        className
      )}
      onClick={onClick}
    >
      {icon && (
        <div className='mb-4'>
          <div className='border-primary inline-flex items-center justify-center rounded-md border p-2'>
            <Icon {...(icon as IIconProps)} />
          </div>
        </div>
      )}
      <div className='space-y-2'>
        <H5>{title}</H5>
        {description && <Paragraph>{description}</Paragraph>}
        {children && <div className='mt-4'>{children}</div>}
      </div>
    </div>
  );
};

export { ElevatedCard };
