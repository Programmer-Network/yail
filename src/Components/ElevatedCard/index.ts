export { ElevatedCard } from "./ElevatedCard";
export * from "./types";
