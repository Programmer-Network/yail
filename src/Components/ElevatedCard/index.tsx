import classNames from "classnames";
import React from "react";

import Icon from "../Icon";
import { IIconProps } from "../Icon/types";
import { H5, Paragraph } from "../Typography";
import { IElevatedCardProps } from "./types";

const ElevatedCard: React.FC<IElevatedCardProps> = ({
  title,
  description,
  icon,
  children,
  className,
  onClick,
  dataTestId = "elevated-card"
}) => {
  return (
    <div
      data-testid={dataTestId}
      className={classNames(
        "yl:flex yl:flex-col yl:rounded-lg yl:p-6 yl:border yl:border-border yl:bg-text/1",
        {
          "yl:cursor-pointer yl:hover:bg-text/3": onClick
        },
        className
      )}
      onClick={onClick}
    >
      {icon && (
        <div className='yl:mb-4'>
          <div className='yl:inline-flex yl:items-center yl:justify-center yl:rounded-md yl:p-2 yl:border yl:border-primary'>
            <Icon {...(icon as IIconProps)} />
          </div>
        </div>
      )}
      <div className='yl:space-y-2'>
        <H5>{title}</H5>
        {description && <Paragraph>{description}</Paragraph>}
        {children && <div className='yl:mt-4'>{children}</div>}
      </div>
    </div>
  );
};

export default ElevatedCard;
