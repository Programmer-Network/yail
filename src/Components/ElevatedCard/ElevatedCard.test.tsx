import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import ElevatedCard from "./";

describe("ElevatedCard component", () => {
  const title = "Curated Resources";
  const description =
    "Explore our collection of tools, libraries, and tutorials to enhance your development workflow.";

  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(
      <ElevatedCard title={title} description={description} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders title and description", () => {
    render(<ElevatedCard title={title} description={description} />);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(description)).toBeInTheDocument();
  });

  test("conditionally renders icon", () => {
    render(
      <ElevatedCard
        title={title}
        description={description}
        icon={{
          iconName: "IconBook",
          dataTestId: "icon-book"
        }}
      />
    );
    expect(screen.getByTestId("icon-book")).toBeInTheDocument();
  });

  test("renders children when provided", () => {
    const childText = "Child Content";
    render(
      <ElevatedCard title={title} description={description}>
        <div data-testid='child-content'>{childText}</div>
      </ElevatedCard>
    );
    expect(screen.getByTestId("child-content")).toBeInTheDocument();
    expect(screen.getByText(childText)).toBeInTheDocument();
  });

  test("handles click events", () => {
    const onClick = vi.fn();
    render(
      <ElevatedCard title={title} description={description} onClick={onClick} />
    );

    fireEvent.click(screen.getByTestId("elevated-card"));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test("applies custom class name", () => {
    const customClass = "custom-card-class";

    render(
      <ElevatedCard
        title={title}
        description={description}
        className={customClass}
      />
    );

    const cardElement = screen.getByTestId("elevated-card");
    expect(cardElement).toHaveClass(customClass);
  });

  test("uses custom data-testid when provided", () => {
    const customTestId = "custom-card-test-id";
    render(
      <ElevatedCard
        title={title}
        description={description}
        dataTestId={customTestId}
      />
    );

    expect(screen.getByTestId(customTestId)).toBeInTheDocument();
  });
});
