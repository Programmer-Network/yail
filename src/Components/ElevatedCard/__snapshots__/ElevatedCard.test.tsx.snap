// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ElevatedCard component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <div
    class="yl:flex yl:flex-col yl:rounded-lg yl:p-6 yl:border yl:border-border yl:bg-text/1"
    data-testid="elevated-card"
  >
    <div
      class="yl:space-y-2"
    >
      <h5
        class="yl:text-md yl:sm:text-lg yl:md:text-xl yl:text-text yl:break-words yl:overflow-hidden yl:font-bold"
      >
        Curated Resources
      </h5>
      <p
        class="yl:text-text yl:break-words yl:overflow-hidden"
      >
        Explore our collection of tools, libraries, and tutorials to enhance your development workflow.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
